### PR TITLE
Added an option to replace disconnected players with AI bots

### PR DESCRIPTION
**Pull request by Mailaender**
A simplified version of https://github.com/OpenRA/OpenRA/pull/8059. Closes https://github.com/OpenRA/OpenRA/issues/3001. Depends on https://github.com/OpenRA/OpenRA/pull/10787.

### CODE REVIEW

File: OpenRA.Game/Settings.cs [hunk 1]
```diff
@@ -73,6 +73,9 @@ public class ServerSettings
 		[Desc("Disallow AI bots.")]
 		public bool LockBots = false;
 
+		[Desc("If a player drops, the AI should take over.")]
+		public bool BotsForDisconnectedClients = true;
```
Comment (pchote) [inline OpenRA.Game/Settings.cs]: This should be a lobby/match option, not a global OpenRA setting.

File: OpenRA.Game/Server/Server.cs [hunk 2]
```diff
@@ -579,7 +580,19 @@ public void DropClient(Connection toDrop, int frame)
...
+					var bot = Map.Rules.Actors["player"].TraitInfos<IBotInfo>().Random(Random);
+					DispatchOrdersToClients(toDrop, 0, new ServerOrder("ReplacePlayerWithBot", bot.Name).Serialize());
+
+					dropClient.IpAddress = null;
+					var admin = LobbyInfo.Clients.First(c => c.IsAdmin);
+					dropClient.BotControllerClientIndex = admin.Index;
+					dropClient.Bot = bot.Name;
+				}
+				else
+					LobbyInfo.Clients.RemoveAll(c => c.Index == toDrop.PlayerIndex);
```
Comment (Phrohdoh) [inline OpenRA.Game/Server/Server.cs]: Is there a reason you aren't using `RemoveAt(toDrop.PlayerIndex)`?

File: OpenRA.Game/Network/UnitOrders.cs [hunk 3]
```diff
@@ -67,6 +67,35 @@ public static void ProcessOrder(OrderManager orderManager, World world, int clie
...
+						var client = orderManager.LobbyInfo.ClientWithIndex(clientId);
+						if (client != null && world != null)
+						{
+							var player = world.FindPlayerByClient(client);
+							if (player == null || player.IsBot)
+								break;
+
+							client.State = Session.ClientState.Ready;
+
+							var admin = orderManager.LobbyInfo.Clients.FirstOrDefault(c => c.IsAdmin);
```
Comment (pchote) [inline OpenRA.Game/Network/UnitOrders.cs]: What if it was the admin who dropped (on a dedicated server)?

Comment (Phrohdoh) [inline OpenRA.Game/Network/UnitOrders.cs]: Would it be possible to sort clients by ping and pick the best result to be the next admin, or is the idea of ping irrelevant if there is no admin/host?

File: OpenRA.Game/Server/Server.cs (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Game/Server/Server.cs]: This line hasn't changed, just moved.

File: OpenRA.Game/Network/UnitOrders.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Game/Network/UnitOrders.cs]: @pchote Then this won't fire. See https://github.com/OpenRA/OpenRA/pull/10502#discussion-diff-49606608R583

@Phrohdoh See https://github.com/OpenRA/OpenRA/pull/10502/files#diff-e34ac0a837f43698d67024b60cd1cf12R598 but I doubt "ping" would be a good measurement.

Comment (pchote) [inline OpenRA.Game/Network/UnitOrders.cs]: That is my point, yes.  Introducing a half-baked feature that works for some clients but not others is asking for trouble.

Comment (pchote) [inline OpenRA.Game/Network/UnitOrders.cs]: Also, if the admin drops then there may also be bots that are now left without control.  Control for all of these should be switched to another player.

**Comment by whinis** [comment]
Played with Nolt and Mailaender, when Nolt left bot took over and started playing as expected. No Desync Observed

File: OpenRA.Game/Network/UnitOrders.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Game/Network/UnitOrders.cs]: Don't try to expand the scope here

> Added an option to replace disconnected players with AI bots

I didn't promise to fix bots on dedicated servers, nor transfer adminship and bot controlling. Those are existing limitations, but I don't think they should be tackled here.

Comment (pchote) [inline OpenRA.Game/Network/UnitOrders.cs]: The pr does not work as advertised:

> Added an option to replace disconnected players with AI bots

Players will not be replaced with an AI bot if they are the server admin.

**Comment by pchote** [comment]
As mentioned above in the inline comments (which may be lost if the pr updates), I feel that this pr has fundamental flaws that cause more problems than they fix.  I have to give a :-1: for that reason.

File: OpenRA.Game/Settings.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Game/Settings.cs]: Dedicated server admins will want to disable this. See the setting above.

Comment (pchote) [inline OpenRA.Game/Settings.cs]: Not all server admins.

File: OpenRA.Game/Network/UnitOrders.cs (hunk 3 above)
Comment (whinis) [inline OpenRA.Game/Network/UnitOrders.cs]: Is there any particular reason the bot controller has to be admin ?
If not couldn't you select a play at random, assign the bots to them and if the dropped player had any bots remake them for different random players ?

**Comment by Micr0Bit** [comment]
- "What if it was the admin who dropped (on a dedicated server)?"
- "Then this won't fire."
- "That is my point, yes. Introducing a half-baked feature that works for some clients but not others is asking for trouble."

this pretty much got me to the point for a :-1:

**Comment by Mailaender** [comment]
Can we please move back to constructive criticism instead of throwing people to the lions with blunt negativity? You are creating a very hostile environment for contributors who give their best to improve the game.

**Comment by pchote** [comment]
Unfortunately that is what happens when the initial review comments pointing out fundamental issues are dismissed with responses like like "Don't try to expand the scope too much" and "I went for KISS here. Don't change half the project, change the GUI for a small feature. Better get it to work now and extend it later. Don't always pull the whole arm, when someone gives you a hand.".   If you're going to shut down the review dialog with a blanket "no, i'm not going to fix that", then there is little choice but to become increasingly blunt in order to get the point across.

**Comment by Mailaender** [comment]
I made a ticket for your request https://github.com/OpenRA/OpenRA/issues/10515. You should have noted that this is a pre-requisite already in https://github.com/OpenRA/OpenRA/pull/8059. Now it feels like an arbitrary request. I also don't have a solution for it and it probably should be implemented, tested and reviewed in it's own pull request.

File: OpenRA.Game/Settings.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Game/Settings.cs]: Is there a reason for a player to not want to have this? The original https://github.com/OpenRA/OpenRA/pull/8059 already had it in so it is easy to copy it, but it felt like a needless switch in the already crowded lobby.

**Comment by pchote** [comment]
The prerequisite #10787 was stalled by desync issues that will probably also be a problem here.  These problems aren't likely to be fixed in the foreseeable future, so unfortunately we can't do much more here either.

**Comment by xp65** [comment]
It would be very cool if this was in the next release, would save many failed games due to a player leaving due to connection timing out later in the game removing all the fun :) Killing an idle player is quote stupid. Adding this feature would be worth a release on its own for me.

**Comment by Micr0Bit** [comment]
@xp65 : i feel you , it seems that the recent increase on new players sadly also increasing the games with people just disconnecting or surrendering too early ... noticeable in ra - teamgames

**Comment by Mailaender** [comment]
If you want this badly, you should first go and verify that https://github.com/OpenRA/OpenRA/pull/10787 works as it should. I did some tests locally with multiple clients and dedicated servers, but that is far from a real field test.

**Comment by pchote** [comment]
Considering the current status of our @OpenRA/engine-hackers it seems unlikely that we will be able to get this into a merge worthy state in any reasonable time frame.  Moving back to the future milestone.

**Comment by Mailaender** [comment]
I fixed this up so that it even works, when the admin disconnects on a dedicated server.

**Comment by pchote** [comment]
See http://www.sleipnirstuff.com/forum/viewtopic.php?t=20232 for related discussion. I expect that enabling this by default will not be recieved well by some players, and considering how toxic that part of our community has become, i’d prefer not to anywhere near this.

**Comment by Mailaender** [comment]
While I don't think you should reward toxic behavior with special attention, the request to disable this in the lobby is valid as it might not be for everyone and players/maps may want to pre-define the AI type.

**Comment by KOYK** [comment]
@pchote "I expect that enabling this by default will not be recieved well by some players"

Yes because their fingers will get tired by clicking on it...I personally find this a good feature and any one who doesn't like it is because they never got a disconnected team mate who is about to fire a nuke in the last minute. By the way, adding staff is good specially if the user can turn them off and on, removing staff in the other hand is a bad thing to do, so toxic or not it doesn't really matter.

As long as it is a lobby option it doesn't matter if you even add chickens instead of tanks to the game.

**Comment by pchote** [comment]
Moving back to the Future milestone alongside #10787.

**Pull request closed without merging**